misc/package/Data Files/MWSE/core/lib/logger/init.lua

```diff
@@ -0,0 +1,714 @@
+            msg = fmt(msg, args[1]())
+        else
+            msg =  fmt(msg, args[1](table.unpack(args, 2)))
+        end
+    else
+        -- nothing was passed as a function, format the message normally
+        msg = fmt(msg, table.unpack(args))
+    end
+    return fmt("[%s] %s", makeHeader(self, record), msg)
+end
+
+
+-- make the logging functions
+for levelStr, level in pairs(LOG_LEVEL) do
+    -- e.g., "DEBUG" -> "debug"
+    ---@param self Logger
+    Logger[string.lower(levelStr)] = function(self, msg, ...)
+        if self.level < level then return end
+
+
+        local str = self:format(self:makeRecord(msg, {...}, level))
+
+        if self.file then
+            self.file:write(str, "\n")
+            self.file:flush()
+        else
+            print(str)
+        end
+    end
+end
+
+-- i am a very good programmer
+Logger.none = nil
+
+-- update `call` to be the same as `debug`. this is so that the line numbers are pulled correctly when using the metamethod.
+---@diagnostic disable-next-line: undefined-field
+logMetatable.__call = Logger.debug
+
+
+function Logger:assert(condition, msg, ...)
+    if condition then return end
+
+    -- cant call `Logger:error` because we need the call to `debug.getinfo` to produce the correct line number. super hacky :/
+    local str = self:format(self:makeRecord(msg, {...}, LOG_LEVEL.ERROR))
+
+    if self.level >= LOG_LEVEL.ERROR then
+        if self.file then
+            self.file:write(str, "\n")
+            self.file:flush()
+        else
+            print(str)
+        end
+    end
+
+    assert(condition, str)
+
+end
+
+function Logger:writeInitializedMessage(version)
+    if self.level < Logger.LEVEL.INFO then return end
+
+    if not version then
+        local metadata = tes3.getLuaModMetadata(self.modDir)
+        if metadata then
+            version = metadata.package.version
+        end
+    end
+    -- need to do it this way so the call to `debug.getinfo` produces the correct line number. super hacky :/
+    local record
+    if version then
+        record = self:makeRecord("Initialized version %s.", {version}, LOG_LEVEL.INFO)
+    else
+        record = self:makeRecord("Mod initialized.", {}, LOG_LEVEL.INFO)
+    end
+
+    local str = self:format(record)
+
+    if self.level >= LOG_LEVEL.ERROR then
+        if self.file then
+            self.file:write(str, "\n")
+            self.file:flush()
+        else
+            print(str)
+        end
+    end
+end
+
+
+
+-- =============================================================================
+-- BACKWARDS COMPATIBILITY
+-- =============================================================================
+
+
+-- support the old way
+---@deprecated 
+---@param levelStr string
+function Logger:setLogLevel(levelStr)
+    local level = LOG_LEVEL[levelStr]
+    if not level then return end
+
+    updateKey(self, "level", level)
+end
+
+
+-- support the old way
+---@deprecated 
+Logger.getLogger = Logger.get
```

Comment:
I am okay with the switch to table argument for `Logger.get`, but `Logger.getLogger` needs to be backwards compatible with 1 string argument.

Reply:
i should have probably clarified this better in the commit, but the `Logger.get` method accepts either a `table` with the appropriate parameters, or a `string` (which is interpretted as the `modName`). so, `Logger.get(modName)` is still valid.

but maybe it's better to have `Logger.get` only accept a `table` argument?